hs/src/Reach/Optimize.hs

```diff
@@ -246,6 +248,99 @@ opt_v2v v = fst <$> opt_v2p v
 opt_v2a :: DLVar -> App DLArg
 opt_v2a v = snd <$> opt_v2p v
 
+class IsCom a where
+  isCom :: a -> Maybe (DLStmt, a)
+
+instance IsCom DLTail where
+  isCom = \case
+    DT_Com m k -> Just (m, k)
+    _ -> Nothing
+
+instance IsCom CTail where
+  isCom = \case
+    CT_Com m k -> Just (m, k)
+    _ -> Nothing
+
+instance IsCom ETail where
+  isCom = \case
+    ET_Com m k -> Just (m, k)
+    _ -> Nothing
+
+instance IsCom LLStep where
+  isCom = \case
+    LLS_Com m k -> Just (m, k)
+    _ -> Nothing
+
+instance IsCom LLConsensus where
+  isCom = \case
+    LLC_Com m k -> Just (m, k)
+    _ -> Nothing
+
+class LiftPure a where
+  -- Returns what `DLVar` is `DL_Set` to and the pure `DLStmt`s along the way
+  -- Returns `Nothing` if there are impure stmts in `a` or there is no `DL_Set`
+  liftPure :: DLVar -> a -> (Maybe DLArg, [DLStmt])
+
+instance LiftPure DLTail where
+  liftPure dv = \case
+    DT_Return _ -> (Nothing, [])
+    DT_Com (DL_Set _ lhs rhs) (DT_Return _)
+      | lhs == dv -> (Just rhs, [])
+    DT_Com ds dt ->
+      case isPure ds of
+        True -> do
+          let (ma, dt') = liftPure dv dt
+          (ma, ds : dt')
+        False -> (Nothing, [])
+
+floatVar :: (DLStmt -> t -> t) -> SrcLoc -> DLVar -> t -> App t
+floatVar mkk at dv k = do
+  eVarsSet <- (liftIO . readIORef) =<< asks eVarsSet
+  case M.lookup dv eVarsSet of
+    -- There were zero branches that could not be lifted, i.e.
+    -- we generated a `const` assign, now drop this `var`.
+    Just 0 -> return $ k
+    _ -> return $ mkk (DL_Var at dv) k
+
+mkVar :: SrcLoc -> DLType -> App DLVar
+mkVar at t = do
+  ctr <- asks eCounter
+  idx <- liftIO $ incCounter ctr
+  return $ DLVar at Nothing t idx
+
+floatIf :: (DLStmt -> b -> b) -> SrcLoc -> DLVar -> DLArg -> DLTail -> DLTail -> b -> App b
+floatIf mkk at ansDv c tt ft k = do
+  case isPure tt && isPure ft of
+    True ->
+      case (liftPure ansDv tt, liftPure ansDv ft) of
+        ((Just ta, t_lifts), (Just fa, f_lifts)) -> do
+          eVarsSetR <- asks eVarsSet
+          eVarsSet <- (liftIO . readIORef) eVarsSetR
+          let mVal = M.lookup ansDv eVarsSet
+          -- Track that we successfully lifted this assignment
+          liftIO $ modifyIORef' eVarsSetR $ M.insert ansDv $ mSub1 mVal
+          let asn v = DL_Let at (DLV_Let DVC_Many v) $ DLE_PrimOp at IF_THEN_ELSE [c, ta, fa]
+          t <- case mVal of
+                Just 1 -> return $ mkk (asn ansDv) k
```

Comment:
This is matching on 1 because it is GOING to be 0 after this? I feel like that's confusing and I want you to match on `mSub1 mVal`